### DISCUSSION TITLE
Use it? Tell us.

### DISCUSSION THREAD

**Discussion by kopf-archiver** [nolar/kopf#100] [Show and tell]
> <a href="https://github.com/nolar"><img align="left" height="50" src="https://avatars0.githubusercontent.com/u/544296?v=4"></a>        An issue by [nolar](https://github.com/nolar) at _2019-06-02 19:51:58+00:00_
> Original URL: https://github.com/zalando-incubator/kopf/issues/100
> &nbsp;

We are happy to know that you like Kopf, and can apply it in your work.

We will be even more happy if you share with us how you use it: operators created, patterns invented, problems solved, so on. Repos? Posts? Demos? Talks? Presentations? — Everything counts.

In additional to the emotional reward for creating a useful tool, this will help us to decide on the feature roadmap, and to adjust Kopf to the real needs of real users.

_For bugs, ideas, feature requests, or just questions, please open a new issue — so that other developers and users can find it later._

---

> <a href="https://github.com/mzizzi"><img align="left" height="30" src="https://avatars0.githubusercontent.com/u/6410431?v=4"></a>            Commented by [mzizzi](https://github.com/mzizzi) at _2019-06-15 23:36:56+00:00_
> &nbsp;

PoC for managing [DataDog](https://www.datadoghq.com/product/) Monitor resources via kube: https://github.com/mzizzi/datadog_operator 

---

> <a href="https://github.com/viruzzo"><img align="left" height="30" src="https://avatars1.githubusercontent.com/u/9071826?v=4"></a>            Commented by [viruzzo](https://github.com/viruzzo) at _2019-08-27 07:13:02+00:00_
> &nbsp;

Bridge for automatically syncing DNS entries for ingresses and services into [MAAS](https://maas.io/) through its API; hacky, but trying to collate a number of different systems together.

---

> <a href="https://github.com/mtparet"><img align="left" height="30" src="https://avatars1.githubusercontent.com/u/754770?v=4"></a>            Commented by [mtparet](https://github.com/mtparet) at _2019-11-24 16:21:10+00:00_
> &nbsp;

I built a first POC in 2 hours to automatically sync secrets from auth0.
Very nice and easy framework/tool!

---

> <a href="https://github.com/janvdvegt"><img align="left" height="30" src="https://avatars3.githubusercontent.com/u/12046878?v=4"></a>            Commented by [janvdvegt](https://github.com/janvdvegt) at _2020-01-26 20:09:39+00:00_
> &nbsp;

We are currently using metacontroller for one of our CRDs but it is deprecated and I did not like some of the choices so for a second CRD I'm trying out kopf. So far I think it looks better. Our system basically wraps code of our users with some of our own code and this code is executed in multiple places in our system. Our first use case will be a wrapper around Jobs that build specific Docker images if not available yet. This way other services can request these asynchronously. The controller will lookup whether these exist already. If not it will create a Job to make and push them, if they exist already it will just put the status to Done. Other services can just interact with these CRD instances then. This also allows us to migrate more easily by creating a bunch of custom resources when we update our own code.

---

> <a href="https://github.com/bukowa"><img align="left" height="30" src="https://avatars1.githubusercontent.com/u/38087302?v=4"></a>            Commented by [bukowa](https://github.com/bukowa) at _2020-01-26 20:51:55+00:00_
> &nbsp;

https://github.com/bukwp/kopfmysql some throw away code but works. The only hard thing to figure out was testing. I found https://github.com/vapor-ware/kubetest nice project but needs some love, i was thinking about some wrapper around K8s py API. Later i tried to create generic dataclasses to define multiple and more complicated crds logic around secret that could be later easy imported into kopf. See https://github.com/bukwp/kopfmysql/tree/v1alpha2/kopfmysql/kopfmysql     i generally love the abstraction and idea of defining crd as a class, like dataclass which already gives nice Type hints that could be used to generate crds in yaml files with methods on the class to handle create update etc sounds briliant to me. Lookup between other resources automatically (see what i tried in secret.py and later mysqlapi) and fields validation based on Type hint could be achieved. Sorry for formatting and errors

---

> <a href="https://github.com/whereismyjetpack"><img align="left" height="30" src="https://avatars2.githubusercontent.com/u/3694886?v=4"></a>            Commented by [whereismyjetpack](https://github.com/whereismyjetpack) at _2020-01-29 03:54:16+00:00_
> &nbsp;

Hi! I am creating an operator that takes a list of keys, generates random passwords, and creates a secret with those generated values. The idea is for containers that read default credentials from a secret, e.g `MYSQL_PASSWORD` I can have the operator generate them, and then inject the secret into each pod without needing to manually generate and provide them

---

> <a href="https://github.com/dahateb"><img align="left" height="30" src="https://avatars2.githubusercontent.com/u/2833350?v=4"></a>            Commented by [dahateb](https://github.com/dahateb) at _2020-03-04 15:55:45+00:00_
> &nbsp;

We built an operator based on kopf to manage the aws eks aws-auth configmap (as decribed here:  https://docs.aws.amazon.com/eks/latest/userguide/add-user-role.html) through custom resources. 

https://github.com/TierMobility/aws-auth-operator

---

> <a href="https://github.com/rvlane"><img align="left" height="30" src="https://avatars3.githubusercontent.com/u/50492970?v=4"></a>            Commented by [rvlane](https://github.com/rvlane) at _2020-03-11 14:59:24+00:00_
> &nbsp;

I was looking at using kopf for our mariadb operator, however we are using a centos8 distribution deploying python 3.6.8.  I assume that since kopf requires python-3.7 that you are using specific feature of that release?  We are tied by the centos distribution so it is unfortunate that kopf isn't supported on python 3.6.  Just looking for confirmation that kopf would not be compatible with python 3.6.  Thanks

---

> <a href="https://github.com/nolar"><img align="left" height="30" src="https://avatars0.githubusercontent.com/u/544296?v=4"></a>            Commented by [nolar](https://github.com/nolar) at _2020-03-11 15:52:04+00:00_
> &nbsp;

[rvlane](https://github.com/rvlane) Yes, you are right. Kopf requires Python 3.7+. See #74 — it was already bound to some of the 3.7 core features a year ago when it was starting and had only few features; and now, Kopf is dependant on many more 3.7+ features, which makes 3.6 support impossible (or highly effortful).

Please consider using pyenv for isolated binary Python 3.7 builds from source, maybe some backported pre-built packages or repositories (note sure how this works in CentOS), or docker containers / k8s pods based on Python 3.7.

---

> <a href="https://github.com/nolar"><img align="left" height="30" src="https://avatars0.githubusercontent.com/u/544296?v=4"></a>            Commented by [nolar](https://github.com/nolar) at _2020-04-21 09:58:21+00:00_
> &nbsp;

[eshepelyuk](https://github.com/eshepelyuk) Can you please create a separate issue for that?

---

> <a href="https://github.com/eshepelyuk"><img align="left" height="30" src="https://avatars3.githubusercontent.com/u/1279477?v=4"></a>            Commented by [eshepelyuk](https://github.com/eshepelyuk) at _2020-04-22 21:46:47+00:00_
> &nbsp;

[nolar](https://github.com/nolar) does kopf have any social communications tools, like gitter, twitter, slack ?

---

> <a href="https://github.com/nolar"><img align="left" height="30" src="https://avatars0.githubusercontent.com/u/544296?v=4"></a>            Commented by [nolar](https://github.com/nolar) at _2020-04-23 10:48:07+00:00_
> &nbsp;

[eshepelyuk](https://github.com/eshepelyuk) No (yet). Twitter [[nolar](https://github.com/nolar)](https://twitter.com/nolar) is one way to reach me. I am also passively present in https://kubernetes.slack.com/ — there is no special chat room for Kopf, so #kubernetes-operators is a good place to start (maybe) — I have instant notifications when someone mentions me.

For feature requests and questions, it is better to use issues here in GitHub — so that the answers can be seen by others in the future; and maybe as a backlog for follow-up improvements based on the developer's feedback.

---

> <a href="https://github.com/corka149"><img align="left" height="30" src="https://avatars2.githubusercontent.com/u/27023359?v=4"></a>            Commented by [corka149](https://github.com/corka149) at _2020-04-29 19:02:49+00:00_
> &nbsp;

With Kopf we implemented an operator. It helps us to control the number of rising systems. This includes tasks like informing people, keeping things update-to-date, keep the system calm.

I am very excited about the coming daemon-decorator.

---

> <a href="https://github.com/asteven"><img align="left" height="30" src="https://avatars0.githubusercontent.com/u/684915?v=4"></a>            Commented by [asteven](https://github.com/asteven) at _2020-05-13 09:57:21+00:00_
> &nbsp;

I'm using kopf to write a dynamic Persistent Volume provisioner for ZFS ([zfs-provisioner](https://github.com/asteven/zfs-provisioner/)).

I am basically rewriting my  [local-zfs-provisioner](https://github.com/asteven/local-zfs-provisioner) from golang to python because the upstream go libs have some issues I can not and don't want to deal with.

Thanks to you, I can code in python instead of in golang.
 \o/

I don't see myself going back ;-)

---

> <a href="https://github.com/thesolution90"><img align="left" height="30" src="https://avatars0.githubusercontent.com/u/46440585?v=4"></a>            Commented by [thesolution90](https://github.com/thesolution90) at _2020-05-20 13:02:31+00:00_
> &nbsp;

I am using kopf for an automated provisioning of secrets and configmaps across namspaces. Therefore I introduced two crds called clusterconfigmap and clustersecret. You can also specify namespace white- or blacklist.
[Repo](https://github.com/genesismining/cluster-config-operator)
A lot of other ideas are in the pipeline for development.

**Comment by cliffburdick** [comment]
We use it for our custom kubernetes scheduler:

https://github.com/Viasat/nhd

**Comment by olivier-mauras** [comment]
Putting my comment from the old repo back here:
We use kopf for a couple of critical controllers.
One is populating - and keeping them as is - namespaces with required configuration like default psp/deployment service accounts, default or custom network policies, etc...
Another one is handling all the resources cleaning - Matching a configuration in a CRD - across our different clusters, from unbounded PV, to review app deployments on dev clusters.
We also have several different other minor ones that are provisioning resources based on different criterias... Most of them are deploy and forget :)

Kopf is de-facto our way to go when we need to implement a custom controller on Kubernetes.

**Comment by chaudum** [comment]
The [CrateDB](https://github.com/crate/crate) Kubernetes Operator which has been open sourced recently is built with Kopf.
https://github.com/crate/crate-operator

**Comment by arkocal** [comment]
[We](https://github.com/Open-IoT-Service-Platform/oisp-services) are experimenting with it to manage Beam/Flink applications on our IoT platform. The main motivation is to enable an administrator to use the platform without expertise in any of the components apart from Kubernetes. Therefore, we do not want them to necessarily know how Flink deployments work, if they can use Kubernetes resources to deploy their applications. 

We preferred Kopf since it is so much less boilerplate than any other framework, our initial implementation was done quite quickly.

**Comment by guhcampos** [comment]
We are Brazil's biggest trading automation platform: https://smarttbot.com and our next-generation robot scheduling is based on Kubernetes and uses `kopf` to manage their lifecycle. We are currently in early alpha but expect our operator to manage several thousand concurrent robots per cluster (with a few additional in-house bred tools)

**Comment by agardnerIT** [comment]
We've built an Operator with kopf which automatically generates OpenTelemetry traces for `Jobs` and `CronJobs`.

https://agardnerit.github.io/tracepusher/usage/k8sjobs/

**Comment by svinther** [comment]
I built a workflow engine based on Kopf that executes DAG-style workflows. It's quite effective for handling asyncio tasks, such as calling various external APIs. Although it's mostly a side project, I'm currently using it for real automation work at several companies I work with.

https://github.com/svinther/fastflow
